### PR TITLE
Fix Issue 16169 - nWayUnion assertion failure

### PR DESCRIPTION
**Pull request by PetruGurita**
Because the BinaryHeap and MultiwayMerge structures did not have proper postblits, whenever these structures were passed by value, resolved in incorrect behaviour with range semantics.
```    
    auto un = nWayUnion(..);
    auto un2 = un;

    while (!un.empty())
    {
        un.front();
        un.popFront();
    }
    while (!un2.empty())
    {
        un2.front();
        un2.popFront();
    }
```
Before, the code from above resulted in assert failure from BinaryHeap. 
In order to correct the problem, a postblit which performs deep copy is defined .
Because the multiWayMerge consumed the range given as parameter and to not break the API,
multiWayMerge is now optionally copyable via the copyable parameter.
This allows for nWayUnion and multiWayMerge to be used in conjunction with chunkBy.

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
I think this would not work in general - ranges can’t be dup-ed.

File: std/algorithm/setops.d [hunk 1]
```diff
@@ -824,14 +824,16 @@ Params:
...
 will leave `ror` modified. Namely, `MultiwayMerge` assumes ownership
 of `ror` and discretionarily swaps and advances elements of it. If
 you want `ror` to preserve its contents after the call, you may
 want to pass a duplicate to `MultiwayMerge` (and perhaps cache the
 duplicate in between calls).
+If you want both the BinaryHeap and the `ror` to be copyable, you may
+instantiate MultiwayMerge with `copyable` as true.
  */
-struct MultiwayMerge(alias less, RangeOfRanges)
+struct MultiwayMerge(alias less, RangeOfRanges, bool copyable = false)
```
[the remark this answers is not in the document]
Comment (PetruGurita) [reply std/algorithm/setops.d]: I see. How to you suggest to solve the issue? By disabling postblit?

**Comment by jercaianu** [comment]
@wilzbach @DmitryOlshansky 
This is clearly broken, should we disable the postblit?
Even though it's a breaking change, code that did assignment of MultiwayMerge is most likely bugged.

**Comment by wilzbach** [comment]
Yeah that sounds like the reasonable thing to do.
An alternative would to rewrite multiwayMerge, s.t. it doesn't assume ownership of `ror`, which I don't think should be too hard (not sure whether we can do this with retaining the same performance too).
OTOH [`merge`](https://dlang.org/phobos/std_algorithm_sorting.html#.merge) managed to do this too with its variadic range overload.

**Comment by jondegenhardt** [comment]
The specific problem reported in issue 16169 should be fixed by PR #6841.

**Comment by jondegenhardt** [comment]
@thewilsonator [Issue 16169](https://issues.dlang.org/show_bug.cgi?id=16169) has been fixed by the merge of PR #6841. (I marked the issue resolved.)

**Comment by jondegenhardt** [comment]
Ping @thewilsonator - I think this PR can be closed. The bug in question was fixed by PR #6841.

**Comment by thewilsonator** [comment]
Thanks for the ping

**Pull request closed without merging**